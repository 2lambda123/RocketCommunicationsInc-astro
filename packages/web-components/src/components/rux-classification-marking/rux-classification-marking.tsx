import { Component, Host, Prop, h, Element, Listen, State } from '@stencil/core'
import { Classification } from '../../common/commonTypes.module'
import { hasSlot } from '../../utils/utils'

/**
 * @part footer-banner - the footer banner
 *
 */
@Component({
    tag: 'rux-classification-marking',
    styleUrl: 'rux-classification-marking.scss',
    shadow: true,
})
export class RuxClassificationMarking {
    @Element() el!: HTMLRuxClassificationMarkingElement
    /**
     * Defines which classification marking will be displayed.
     */
    @Prop({ reflect: true }) classification: Classification = 'unclassified'
    /**
     * Allows additional text labels to be added to the marking
     */
    @Prop() label?: string
    /**
     * Declares the marking as a `tag` rather than the default banner style
     */
    @Prop({ reflect: true }) tag: boolean = false

    @State() isWrapper: boolean = hasSlot(this.el)

    @Listen('slotchange')
    private _handleSlotChange() {
        this.isWrapper = hasSlot(this.el)
    }

    connectedCallback() {
        this._handleSlotChange = this._handleSlotChange.bind(this)
    }

    disconnectedCallback() {
        this.el!.shadowRoot!.removeEventListener(
            'slotchange',
            this._handleSlotChange
        )
    }

    get type(): 'tag' | 'banner' {
        return this.tag ? 'tag' : 'banner'
    }

    private _getDisplayData(): string {
        const markings = {
            banner: {
                cui: 'cui',
                controlled: 'controlled',
                confidential: 'confidential',
                secret: 'secret',
                'top-secret': 'top secret',
                'top-secret-sci': 'top secret//sci',
                unclassified: 'unclassified',
            },
            tag: {
                cui: 'cui',
                controlled: 'cui',
                confidential: 'c',
                secret: 's',
                'top-secret': 'ts',
                'top-secret-sci': 'TS//SCI',
                unclassified: 'u',
            },
        }

        const text = markings[this.type][this.classification]
        return text
    }

    render() {
        const { isWrapper, label, tag, type } = this
        return (
            <Host>
                <div
                    class={{
                        'rux-classification': true,
                        'rux-classification--tag': type === 'tag',
                        'rux-classification--banner': type === 'banner',
                    }}
                >
                    {this._getDisplayData()}
                    {label}
                </div>
                <slot></slot>
                {isWrapper && !tag && (
                    <div
                        class={{
                            'rux-classification': true,
                            'rux-classification--banner': type === 'banner',
                            'rux-classification--banner__footer':
                                isWrapper === true,
                        }}
                        part="footer-banner"
                    >
                        {this._getDisplayData()}
                        {label}
                    </div>
                )}
            </Host>
        )
    }
}
