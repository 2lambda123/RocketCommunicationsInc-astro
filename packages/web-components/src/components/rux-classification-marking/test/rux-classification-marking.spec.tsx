import { newSpecPage } from '@stencil/core/testing'
import { RuxClassificationMarking } from '../rux-classification-marking'

describe('rux-classification-marking', () => {
    it('builds', async () => {
        const classificationMarking = new RuxClassificationMarking()

        expect(classificationMarking).toBeTruthy()
        expect(classificationMarking).toEqual({
            classification: 'unclassified',
            isWrapper: false,
            tag: false,
        })
    })

    it('renders component with no properties set', async () => {
        const page = await newSpecPage({
            components: [RuxClassificationMarking],
            html: `<rux-classification-marking></rux-classification-marking>`,
        })
        expect(page.root).toMatchSnapshot()
    })

    describe('rux-classification-marking banners', () => {
        it('renders unclassified banner', async () => {
            const page = await newSpecPage({
                components: [RuxClassificationMarking],
                html: `<rux-classification-marking classification="unclassified"></rux-classification-marking>`,
            })
            expect(page.root).toMatchSnapshot()
        })

        it('renders controlled banner', async () => {
            const page = await newSpecPage({
                components: [RuxClassificationMarking],
                html: `<rux-classification-marking classification="controlled"></rux-classification-marking>`,
            })
            expect(page.root).toMatchSnapshot()
        })

        it('renders cui banner', async () => {
            const page = await newSpecPage({
                components: [RuxClassificationMarking],
                html: `<rux-classification-marking classification="cui"></rux-classification-marking>`,
            })
            expect(page.root).toMatchSnapshot()
        })

        it('renders confidential banner', async () => {
            const page = await newSpecPage({
                components: [RuxClassificationMarking],
                html: `<rux-classification-marking classification="confidential"></rux-classification-marking>`,
            })
            expect(page.root).toMatchSnapshot()
        })

        it('renders secret banner', async () => {
            const page = await newSpecPage({
                components: [RuxClassificationMarking],
                html: `<rux-classification-marking classification="secret"></rux-classification-marking>`,
            })
            expect(page.root).toMatchSnapshot()
        })

        it('renders top secret banner', async () => {
            const page = await newSpecPage({
                components: [RuxClassificationMarking],
                html: `<rux-classification-marking classification="top-secret"></rux-classification-marking>`,
            })
            expect(page.root).toMatchSnapshot()
        })

        it('renders top secret sci banner', async () => {
            const page = await newSpecPage({
                components: [RuxClassificationMarking],
                html: `<rux-classification-marking classification="top-secret-sci"></rux-classification-marking>`,
            })
            expect(page.root).toMatchSnapshot()
        })

        it('renders tincorrect option banner', async () => {
            const page = await newSpecPage({
                components: [RuxClassificationMarking],
                html: `<rux-classification-marking classification="notapprovedoption"></rux-classification-marking>`,
            })
            expect(page.root).toMatchSnapshot()
        })
    })

    describe('rux-classification-marking tags', () => {
        it('renders unclassified tag', async () => {
            const page = await newSpecPage({
                components: [RuxClassificationMarking],
                html: `<rux-classification-marking classification="unclassified" tag=true></rux-classification-marking>`,
            })
            expect(page.root).toMatchSnapshot()
        })

        it('renders controlled tag', async () => {
            const page = await newSpecPage({
                components: [RuxClassificationMarking],
                html: `<rux-classification-marking classification="controlled" tag=true></rux-classification-marking>`,
            })
            expect(page.root).toMatchSnapshot()
        })

        it('renders confidential tag', async () => {
            const page = await newSpecPage({
                components: [RuxClassificationMarking],
                html: `<rux-classification-marking classification="confidential" tag=true></rux-classification-marking>`,
            })
            expect(page.root).toMatchSnapshot()
        })

        it('renders secret tag', async () => {
            const page = await newSpecPage({
                components: [RuxClassificationMarking],
                html: `<rux-classification-marking classification="secret" tag=true></rux-classification-marking>`,
            })
            expect(page.root).toMatchSnapshot()
        })

        it('renders top secret tag', async () => {
            const page = await newSpecPage({
                components: [RuxClassificationMarking],
                html: `<rux-classification-marking classification="top-secret" tag=true></rux-classification-marking>`,
            })
            expect(page.root).toMatchSnapshot()
        })

        it('renders top secret sci tag', async () => {
            const page = await newSpecPage({
                components: [RuxClassificationMarking],
                html: `<rux-classification-marking classification="top-secret-sci" tag=true></rux-classification-marking>`,
            })
            expect(page.root).toMatchSnapshot()
        })

        it('renders incorrect option tag', async () => {
            const page = await newSpecPage({
                components: [RuxClassificationMarking],
                html: `<rux-classification-marking classification="notapprovedoption" tag=true></rux-classification-marking>`,
            })
            expect(page.root).toMatchSnapshot()
        })
    })

    describe('rux-classification-marking label', () => {
        it('renders unclassified banner label', async () => {
            const page = await newSpecPage({
                components: [RuxClassificationMarking],
                html: `<rux-classification-marking classification="unclassified" label="-hello world-"></rux-classification-marking>`,
            })
            expect(page.root).toMatchSnapshot()
        })

        it('renders unclassified tag label', async () => {
            const page = await newSpecPage({
                components: [RuxClassificationMarking],
                html: `<rux-classification-marking classification="unclassified" tag=true label="-hello world-"></rux-classification-marking>`,
            })
            expect(page.root).toMatchSnapshot()
        })

        //This test does not test for the proper expectation but excludes the second footer-banner div due to an unknown bug in jest
        it('renders footer banner', async () => {
            const page = await newSpecPage({
                components: [RuxClassificationMarking],
                html: `
          <rux-classification-marking classification="secret">
            <h1>Test title for footer banner</h1>
          </rux-classification-marking>
        `,
            })

            expect(page.root).toMatchSnapshot()
        })
    })
})
