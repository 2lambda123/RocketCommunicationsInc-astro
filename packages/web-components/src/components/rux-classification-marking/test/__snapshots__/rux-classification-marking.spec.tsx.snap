// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rux-classification-marking renders component with no properties set 1`] = `
<rux-classification-marking classification="unclassified">
  <mock:shadow-root>
    <div class="rux-classification rux-classification--banner" part="tag header">
      unclassified
    </div>
    <slot></slot>
  </mock:shadow-root>
</rux-classification-marking>
`;

exports[`rux-classification-marking rux-classification-marking banners renders confidential banner 1`] = `
<rux-classification-marking classification="confidential">
  <mock:shadow-root>
    <div class="rux-classification rux-classification--banner" part="tag header">
      confidential
    </div>
    <slot></slot>
  </mock:shadow-root>
</rux-classification-marking>
`;

exports[`rux-classification-marking rux-classification-marking banners renders controlled banner 1`] = `
<rux-classification-marking classification="controlled">
  <mock:shadow-root>
    <div class="rux-classification rux-classification--banner" part="tag header">
      controlled
    </div>
    <slot></slot>
  </mock:shadow-root>
</rux-classification-marking>
`;

exports[`rux-classification-marking rux-classification-marking banners renders cui banner 1`] = `
<rux-classification-marking classification="cui">
  <mock:shadow-root>
    <div class="rux-classification rux-classification--banner" part="tag header">
      cui
    </div>
    <slot></slot>
  </mock:shadow-root>
</rux-classification-marking>
`;

exports[`rux-classification-marking rux-classification-marking banners renders secret banner 1`] = `
<rux-classification-marking classification="secret">
  <mock:shadow-root>
    <div class="rux-classification rux-classification--banner" part="tag header">
      secret
    </div>
    <slot></slot>
  </mock:shadow-root>
</rux-classification-marking>
`;

exports[`rux-classification-marking rux-classification-marking banners renders tincorrect option banner 1`] = `
<rux-classification-marking classification="notapprovedoption">
  <mock:shadow-root>
    <div class="rux-classification rux-classification--banner" part="tag header">
      unclassified
    </div>
  <slot></slot>
  </mock:shadow-root>
</rux-classification-marking>
`;

exports[`rux-classification-marking rux-classification-marking banners renders top secret banner 1`] = `
<rux-classification-marking classification="top-secret">
  <mock:shadow-root>
    <div class="rux-classification rux-classification--banner" part="tag header">
      top secret
    </div>
    <slot></slot>
  </mock:shadow-root>
</rux-classification-marking>
`;

exports[`rux-classification-marking rux-classification-marking banners renders top secret sci banner 1`] = `
<rux-classification-marking classification="top-secret-sci">
  <mock:shadow-root>
    <div class="rux-classification rux-classification--banner" part="tag header">
      top secret//sci
    </div>
    <slot></slot>
  </mock:shadow-root>
</rux-classification-marking>
`;

exports[`rux-classification-marking rux-classification-marking banners renders unclassified banner 1`] = `
<rux-classification-marking classification="unclassified">
  <mock:shadow-root>
    <div class="rux-classification rux-classification--banner" part="tag header">
      unclassified
    </div>
    <slot></slot>
  </mock:shadow-root>
</rux-classification-marking>
`;

exports[`rux-classification-marking rux-classification-marking label renders footer banner 1`] = `
<rux-classification-marking classification="secret">
  <mock:shadow-root>
    <div class="rux-classification rux-classification--banner" part="tag header">
      secret
    </div>
    <slot></slot>
  </mock:shadow-root>
  <h1>
    Test title for footer banner
  </h1>
</rux-classification-marking>
`;

exports[`rux-classification-marking rux-classification-marking label renders unclassified banner label 1`] = `
<rux-classification-marking classification="unclassified" label="-hello world-">
  <mock:shadow-root>
    <div class="rux-classification rux-classification--banner" part="tag header">
      unclassified-hello world-
    </div>
    <slot></slot>
  </mock:shadow-root>
</rux-classification-marking>
`;

exports[`rux-classification-marking rux-classification-marking label renders unclassified tag label 1`] = `
<rux-classification-marking classification="unclassified" label="-hello world-" tag="">
  <mock:shadow-root>
    <div class="rux-classification rux-classification--tag" part="tag header">
      u-hello world-
    </div>
    <slot></slot>
  </mock:shadow-root>
</rux-classification-marking>
`;

exports[`rux-classification-marking rux-classification-marking tags renders confidential tag 1`] = `
<rux-classification-marking classification="confidential" tag="">
  <mock:shadow-root>
    <div class="rux-classification rux-classification--tag" part="tag header">
      c
    </div>
    <slot></slot>
  </mock:shadow-root>
</rux-classification-marking>
`;

exports[`rux-classification-marking rux-classification-marking tags renders controlled tag 1`] = `
<rux-classification-marking classification="controlled" tag="">
  <mock:shadow-root>
    <div class="rux-classification rux-classification--tag" part="tag header">
      cui
    </div>
    <slot></slot>
  </mock:shadow-root>
</rux-classification-marking>
`;

exports[`rux-classification-marking rux-classification-marking tags renders incorrect option tag 1`] = `
<rux-classification-marking classification="notapprovedoption" tag="">
  <mock:shadow-root>
    <div class="rux-classification rux-classification--tag" part="tag header">
      u
    </div>
    <slot></slot>
  </mock:shadow-root>
</rux-classification-marking>
`;

exports[`rux-classification-marking rux-classification-marking tags renders secret tag 1`] = `
<rux-classification-marking classification="secret" tag="">
  <mock:shadow-root>
    <div class="rux-classification rux-classification--tag" part="tag header">
      s
    </div>
    <slot></slot>
  </mock:shadow-root>
</rux-classification-marking>
`;

exports[`rux-classification-marking rux-classification-marking tags renders top secret sci tag 1`] = `
<rux-classification-marking classification="top-secret-sci" tag="">
  <mock:shadow-root>
    <div class="rux-classification rux-classification--tag" part="tag header">
      TS//SCI
    </div>
    <slot></slot>
  </mock:shadow-root>
</rux-classification-marking>
`;

exports[`rux-classification-marking rux-classification-marking tags renders top secret tag 1`] = `
<rux-classification-marking classification="top-secret" tag="">
  <mock:shadow-root>
    <div class="rux-classification rux-classification--tag" part="tag header">
      ts
    </div>
    <slot></slot>
  </mock:shadow-root>
</rux-classification-marking>
`;

exports[`rux-classification-marking rux-classification-marking tags renders unclassified tag 1`] = `
<rux-classification-marking classification="unclassified" tag="">
  <mock:shadow-root>
    <div class="rux-classification rux-classification--tag" part="tag header">
      u
    </div>
    <slot></slot>
  </mock:shadow-root>
</rux-classification-marking>
`;
